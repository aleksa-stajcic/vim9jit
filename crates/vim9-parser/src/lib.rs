#![allow(dead_code)]

use anyhow::Result;
use vim9_lexer::new_lexer;
use vim9_lexer::Lexer;
use vim9_lexer::Token;
use vim9_lexer::TokenKind;

#[derive(Debug)]
pub struct Program {
    commands: Vec<ExCommand>,
}

#[derive(Debug, PartialEq)]
pub enum ExCommand {
    Vim9Script(Vim9ScriptCommand),
    Var(VarCommand),
    Return(ReturnCommand),
    Def(DefCommand),
    If(IfCommand),
    Call(CallCommand),
    Finish(Token),

    Skip,
    EndOfFile,
    Comment(Token),
    NoOp(Token),
}

#[derive(Debug, PartialEq)]
pub struct Vim9ScriptCommand {
    noclear: bool,
    eol: Token,
}

#[derive(Debug, PartialEq)]
pub struct VarCommand {
    var: Token,
    pub name: Name,
    equal: Token,
    pub expr: Expression,
    eol: Token,
}

impl VarCommand {
    pub fn parse(parser: &mut Parser) -> Result<ExCommand> {
        let var = VarCommand {
            var: parser.expect_token_with_text(TokenKind::Identifier, "var")?,
            name: Name::parse(parser)?,
            // TODO: Type Hints
            equal: parser.expect_token(TokenKind::Equal)?,
            expr: Expression::parse(parser)?,
            eol: parser.expect_eol()?,
        };

        Ok(ExCommand::Var(var))
    }
}

#[derive(Debug, PartialEq)]
pub struct Body {
    pub commands: Vec<ExCommand>,
}

impl Body {
    pub fn parse_until(parser: &mut Parser, identifier: &str) -> Result<Body> {
        let mut commands = vec![];
        while parser.current_token.text != identifier {
            commands.push(parser.parse_command()?);
        }

        Ok(Body { commands })
    }
}

#[derive(Debug, PartialEq)]
pub struct IfCommand {
    if_tok: Token,
    pub condition: Expression,
    if_eol: Token,
    pub body: Body,
    endif_tok: Token,
    endif_eol: Token,
}

impl IfCommand {
    pub fn parser(parser: &mut Parser) -> Result<ExCommand> {
        Ok(ExCommand::If(IfCommand {
            if_tok: parser.expect_identifier_with_text("if")?,
            condition: Expression::parse(parser)?,
            if_eol: parser.expect_eol()?,
            body: Body::parse_until(parser, "endif")?,
            endif_tok: parser.expect_identifier_with_text("endif")?,
            endif_eol: parser.expect_eol()?,
        }))
    }
}

#[derive(Debug, PartialEq)]
pub struct DefCommand {
    def: Token,
    pub name: Name,
    pub args: Signature,
    pub ret: Option<Type>,
    def_eol: Token,
    pub body: Body,
    enddef: Token,
    end_eol: Token,
}

impl DefCommand {
    pub fn parse(parser: &mut Parser) -> Result<ExCommand> {
        Ok(ExCommand::Def(DefCommand {
            def: parser.expect_identifier_with_text("def")?,
            name: Name::parse(parser)?,
            args: Signature::parse(parser)?,
            ret: {
                if parser.current_token.kind == TokenKind::Colon {
                    Some(Type {
                        colon: parser.expect_token(TokenKind::Colon)?,
                        name: Name::parse(parser)?,
                    })
                } else {
                    None
                }
            },
            def_eol: parser.expect_eol()?,
            body: Body::parse_until(parser, "enddef")?,
            enddef: parser.expect_identifier_with_text("enddef")?,
            end_eol: parser.expect_eol()?,
        }))
    }
}

#[derive(Debug, PartialEq)]
pub struct CallCommand {
    call: Option<Token>,
    pub name: Name,
    pub args: Arguments,
    eol: Token,
}

impl CallCommand {
    pub fn parse(parser: &mut Parser) -> Result<ExCommand> {
        Ok(ExCommand::Call(CallCommand {
            call: None,
            name: Name::parse(parser)?,
            args: Arguments::parse(parser)?,
            eol: parser.expect_eol()?,
        }))
    }
}

#[derive(Debug, PartialEq)]
pub struct Arguments {
    open: Token,
    pub args: Vec<Expression>,
    close: Token,
}

impl Arguments {
    pub fn parse(parser: &mut Parser) -> Result<Arguments> {
        Ok(Arguments {
            open: parser.expect_token(TokenKind::LeftParen)?,
            args: {
                let mut args = vec![];
                while parser.current_token.kind != TokenKind::RightParen {
                    args.push(Expression::parse(parser)?);

                    // TODO: Consume ','?
                }

                args
            },
            close: parser.expect_token(TokenKind::RightParen)?,
        })
    }
}

#[derive(Debug, PartialEq)]
pub struct Type {
    colon: Token,
    pub name: Name,
}

#[derive(Debug, PartialEq)]
pub struct Signature {
    open: Token,
    close: Token,
}

impl Signature {
    fn parse(parser: &mut Parser) -> Result<Signature> {
        Ok(Self {
            open: parser.expect_token(TokenKind::LeftParen)?,
            close: parser.expect_token(TokenKind::RightParen)?,
        })
    }
}

#[derive(Debug, PartialEq)]
pub struct Parameter {
    pub name: Name,
    colon: Token,
    // pub typ: ,
}

#[derive(Debug, PartialEq)]
pub struct ReturnCommand {
    ret: Token,
    pub expr: Expression,
    eol: Token,
}

impl ReturnCommand {
    pub fn parse(parser: &mut Parser) -> Result<ExCommand> {
        Ok(ExCommand::Return(Self {
            ret: parser.expect_token_with_text(TokenKind::Identifier, "return")?,
            expr: Expression::parse(parser)?,
            eol: parser.expect_eol()?,
        }))
    }
}

#[derive(Debug, PartialEq)]
pub enum Name {
    Identifier(RawIdentifier),
    ScopedIdentifier,
}

impl Name {
    fn parse(parser: &mut Parser) -> Result<Name> {
        // Todo: Other names
        Ok(Name::Identifier(RawIdentifier {
            name: parser.expect_token(TokenKind::Identifier)?.text,
        }))
    }
}

#[derive(Debug, PartialEq)]
pub enum Expression {
    Name,
    Number(Number),
    String(VimString),
}

impl Expression {
    pub fn parse(parser: &mut Parser) -> Result<Expression> {
        let expr = match &parser.current_token.kind {
            TokenKind::Integer => Expression::Number(Number {
                value: parser.current_token.text.clone(),
            }),
            TokenKind::DoubleQuoteString => {
                Expression::String(VimString::DoubleQuote(parser.current_token.text.clone()))
            }
            TokenKind::SingleQuoteString => {
                Expression::String(VimString::SingleQuote(parser.current_token.text.clone()))
            }

            _ => todo!("expresson parse: {:?}", parser.current_token),
        };

        parser.next_token();
        Ok(expr)
    }
}

#[derive(Debug, PartialEq)]
pub enum VimString {
    SingleQuote(String),
    DoubleQuote(String),
}

#[derive(Debug, PartialEq)]
pub struct RawIdentifier {
    name: String,
}

#[derive(Debug, PartialEq)]
pub struct Number {
    // ??? should this be a number?
    value: String,
}

#[derive(Debug)]
pub struct Parser {
    lexer: Lexer,
    current_token: Token,
    peek_token: Token,
}

impl Parser {
    pub fn new(mut lexer: Lexer) -> Self {
        Self {
            current_token: lexer.next_token(),
            peek_token: lexer.next_token(),
            lexer,
        }
    }

    pub fn expect_eol(&mut self) -> Result<Token> {
        self.expect_token(TokenKind::EndOfLine)
    }

    pub fn expect_token(&mut self, kind: TokenKind) -> Result<Token> {
        let token = self.current_token.clone();
        if token.kind != kind {
            return Err(anyhow::anyhow!("Got token: {:?}, Expected: {:?}", token, kind));
        }

        self.next_token();
        Ok(token)
    }

    pub fn expect_token_with_text(&mut self, kind: TokenKind, text: &str) -> Result<Token> {
        let token = self.current_token.clone();
        if token.kind != kind {
            panic!("Got token: {:?}, Expected kind: {:?}", token, kind);
            return Err(anyhow::anyhow!("Got token: {:?}, Expected kind: {:?}", token, kind));
        }

        if token.text != text {
            panic!("Got token: {:?}, Expected text: {:?}", token, text);
            return Err(anyhow::anyhow!("Got token: {:?}, Expected text: {:?}", token, text));
        }

        self.next_token();
        Ok(token)
    }

    pub fn peek_identifier_with_text(&self, text: &str) -> bool {
        self.peek_token.kind == TokenKind::Identifier && self.peek_token.text == text
    }

    pub fn expect_identifier_with_text(&mut self, text: &str) -> Result<Token> {
        self.expect_token_with_text(TokenKind::Identifier, text)
    }

    pub fn next_token(&mut self) -> Token {
        self.current_token = self.peek_token.clone();
        self.peek_token = self.lexer.next_token();

        self.current_token.clone()
    }

    fn command_match(&self, full: &str) -> bool {
        self.current_token.text == full
    }

    pub fn parse_command(&mut self) -> Result<ExCommand> {
        // For the following branches, you need to return early if it completely consumes
        // the last character and advances past.
        //
        // This is the desired behavior for `parse` which will consume until the end of line
        // generally speaking.
        let ex = match &self.current_token.kind {
            TokenKind::EndOfFile => panic!("EndOfFile!"),
            TokenKind::Comment => ExCommand::Comment(self.current_token.clone()),
            TokenKind::EndOfLine => ExCommand::NoOp(self.current_token.clone()),
            TokenKind::Identifier => {
                if self.command_match("vim9script") {
                    self.next_token();
                    ExCommand::Vim9Script(Vim9ScriptCommand {
                        noclear: if self.current_token.kind == TokenKind::EndOfLine {
                            false
                        } else {
                            self.expect_identifier_with_text("noclear")?;
                            true
                        },
                        eol: self.expect_eol()?,
                    })
                } else if self.command_match("var") {
                    return Ok(VarCommand::parse(self)?);
                } else if self.command_match("def") {
                    return Ok(DefCommand::parse(self)?);
                } else if self.command_match("return") {
                    return Ok(ReturnCommand::parse(self)?);
                } else if self.command_match("if") {
                    return Ok(IfCommand::parser(self)?);
                } else if self.command_match("finish") {
                    ExCommand::Finish(self.current_token.clone())
                } else {
                    if self.peek_token.kind == TokenKind::LeftParen {
                        return Ok(CallCommand::parse(self)?);
                    } else {
                        ExCommand::NoOp(self.current_token.clone())
                    }
                }
            }
            _ => ExCommand::NoOp(self.current_token.clone()),
        };

        self.next_token();
        Ok(ex)
    }

    pub fn parse_program(&mut self) -> Program {
        let mut program = Program { commands: vec![] };

        while self.current_token.kind != TokenKind::EndOfFile {
            let command = self.parse_command().unwrap();
            if command != ExCommand::Skip {
                program.commands.push(command);
            }
        }

        program
    }
}

fn snapshot_parsing(input: &str) -> String {
    let lexer = new_lexer(input);
    let mut parser = Parser::new(lexer);
    let program = parser.parse_program();

    format!("{:#?}", program.commands)
}

#[cfg(test)]
mod test {
    use super::*;

    macro_rules! snapshot {
        ($name:tt, $path:tt) => {
            #[test]
            fn $name() {
                let contents = include_str!($path);
                let mut settings = insta::Settings::clone_current();
                settings.set_snapshot_path("../testdata/output/");
                settings.bind(|| {
                    insta::assert_snapshot!(snapshot_parsing(contents));
                });
            }
        };
    }

    snapshot!(test_var, "../testdata/snapshots/simple_var.vim");
    snapshot!(test_ret, "../testdata/snapshots/simple_ret.vim");
    snapshot!(test_comment, "../testdata/snapshots/comment.vim");
    snapshot!(test_header, "../testdata/snapshots/header.vim");

    // TODO: Slowly but surely, we can work towards this
    // snapshot!(test_matchparen, "../testdata/snapshots/matchparen.vim");
}
