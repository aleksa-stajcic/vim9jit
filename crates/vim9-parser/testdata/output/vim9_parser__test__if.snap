---
source: crates/vim9-parser/src/lib.rs
assertion_line: 2171
expression: snapshot_parsing(contents)
---
[
    Vim9Script(
        Vim9ScriptCommand {
            noclear: false,
            eol: Token(EndOfLine, "\n", (0,10)->(0,10)),
        },
    ),
    Comment(
        Token(Comment, "# multiple expressions", (2,0)->(2,22)),
    ),
    If(
        IfCommand {
            if_tok: Token(Identifier, "if", (3,0)->(3,2)),
            condition: Infix(
                InfixExpression {
                    token: Token(Or, "||", (3,16)->(3,18)),
                    operator: Or,
                    left: Call(
                        f: Identifier(Raw(pumvisible)) arg: [],
                    ),
                    right: Grouped(
                        GroupedExpression {
                            open: Token(LeftParen, "(", (3,19)->(3,20)),
                            expr: Infix(
                                InfixExpression {
                                    token: Token(And, "&&", (3,38)->(3,40)),
                                    operator: And,
                                    left: Infix(
                                        InfixExpression {
                                            token: Token(LessThan, "<", (3,34)->(3,35)),
                                            operator: LessThan,
                                            left: Call(
                                                f: Identifier(Raw(str2nr)) arg: [
                                                    VimOption(
                                                        VimOption {
                                                            ampersand: Token(Ampersand, "&", (3,27)->(3,28)),
                                                            option: Literal {
                                                                token: Token(Identifier, "t_Co", (3,28)->(3,32)),
                                                            },
                                                        },
                                                    ),
                                                ],
                                            ),
                                            right: Number(
                                                VimNumber {
                                                    value: "8",
                                                },
                                            ),
                                        },
                                    ),
                                    right: Prefix(
                                        PrefixExpression {
                                            token: Token(Bang, "!", (3,41)->(3,42)),
                                            operator: Bang,
                                            right: Call(
                                                f: Identifier(Raw(has)) arg: [
                                                    String(
                                                        SingleQuote(
                                                            "gui_running",
                                                        ),
                                                    ),
                                                ],
                                            ),
                                        },
                                    ),
                                },
                            ),
                            close: Token(RightParen, ")", (3,60)->(3,61)),
                        },
                    ),
                },
            ),
            if_eol: Token(EndOfLine, "\n", (3,61)->(3,61)),
            body: Body {
                commands: [
                    Finish(
                        FinishCommand {
                            finish: Token(Identifier, "finish", (4,2)->(4,8)),
                            eol: Token(EndOfLine, "\n", (4,8)->(4,8)),
                        },
                    ),
                ],
            },
            elseifs: [],
            else_command: None,
            endif_tok: Token(Identifier, "endif", (5,0)->(5,5)),
            endif_eol: Token(EndOfLine, "\n", (5,5)->(5,5)),
        },
    ),
    NoOp(
        Token(EndOfLine, "\n", (6,0)->(6,0)),
    ),
    Var(
        VarCommand {
            var: Token(Identifier, "var", (7,0)->(7,3)),
            ty: None,
            name: Raw(x),
            equal: Token(Equal, "=", (7,6)->(7,7)),
            expr: Number(
                VimNumber {
                    value: "20",
                },
            ),
            eol: Token(EndOfLine, "\n", (7,10)->(7,10)),
        },
    ),
    If(
        IfCommand {
            if_tok: Token(Identifier, "if", (8,0)->(8,2)),
            condition: Infix(
                InfixExpression {
                    token: Token(LessThan, "<", (8,5)->(8,6)),
                    operator: LessThan,
                    left: Identifier(
                        Raw(x),
                    ),
                    right: Number(
                        VimNumber {
                            value: "10",
                        },
                    ),
                },
            ),
            if_eol: Token(EndOfLine, "\n", (8,9)->(8,9)),
            body: Body {
                commands: [
                    Echo(
                        EchoCommand {
                            echo: Token(Identifier, "echo", (9,2)->(9,6)),
                            expr: String(
                                DoubleQuote(
                                    "10",
                                ),
                            ),
                            eol: Token(EndOfLine, "\n", (9,11)->(9,11)),
                        },
                    ),
                ],
            },
            elseifs: [
                ElseIfCommand {
                    elseif_tok: Token(Identifier, "elseif", (10,0)->(10,6)),
                    condition: Infix(
                        InfixExpression {
                            token: Token(LessThan, "<", (10,9)->(10,10)),
                            operator: LessThan,
                            left: Identifier(
                                Raw(x),
                            ),
                            right: Number(
                                VimNumber {
                                    value: "15",
                                },
                            ),
                        },
                    ),
                    elseif_eol: Token(EndOfLine, "\n", (10,13)->(10,13)),
                    body: Body {
                        commands: [
                            Echo(
                                EchoCommand {
                                    echo: Token(Identifier, "echo", (11,2)->(11,6)),
                                    expr: String(
                                        DoubleQuote(
                                            "15",
                                        ),
                                    ),
                                    eol: Token(EndOfLine, "\n", (11,11)->(11,11)),
                                },
                            ),
                        ],
                    },
                },
                ElseIfCommand {
                    elseif_tok: Token(Identifier, "elseif", (12,0)->(12,6)),
                    condition: Infix(
                        InfixExpression {
                            token: Token(LessThan, "<", (12,9)->(12,10)),
                            operator: LessThan,
                            left: Identifier(
                                Raw(x),
                            ),
                            right: Number(
                                VimNumber {
                                    value: "30",
                                },
                            ),
                        },
                    ),
                    elseif_eol: Token(EndOfLine, "\n", (12,13)->(12,13)),
                    body: Body {
                        commands: [
                            Echo(
                                EchoCommand {
                                    echo: Token(Identifier, "echo", (13,2)->(13,6)),
                                    expr: String(
                                        DoubleQuote(
                                            "30",
                                        ),
                                    ),
                                    eol: Token(EndOfLine, "\n", (13,11)->(13,11)),
                                },
                            ),
                        ],
                    },
                },
            ],
            else_command: Some(
                ElseCommand {
                    else_tok: Token(Identifier, "else", (14,0)->(14,4)),
                    else_eol: Token(EndOfLine, "\n", (14,4)->(14,4)),
                    body: Body {
                        commands: [
                            Echo(
                                EchoCommand {
                                    echo: Token(Identifier, "echo", (15,2)->(15,6)),
                                    expr: String(
                                        DoubleQuote(
                                            "else",
                                        ),
                                    ),
                                    eol: Token(EndOfLine, "\n", (15,13)->(15,13)),
                                },
                            ),
                        ],
                    },
                },
            ),
            endif_tok: Token(Identifier, "endif", (16,0)->(16,5)),
            endif_eol: Token(EndOfLine, "\n", (16,5)->(16,5)),
        },
    ),
]
